### ISSUE TITLE
Lists with newlines between the items are malformed

### ISSUE THREAD

**Issue by tomkelsey** [thibaudcolas/draftjs-filters#65]
**Do you want to request a feature or report a _bug_?**

Bug

**What is the current behavior?**

Lists with newlines between the items are malformed

**If the current behavior is a bug, please provide the steps to reproduce and if possible a minimal demo of the problem. GIFs and screenshots are very helpful too.**

On entering text such as (NB: copy/pasting from below didn't work - I had to type it)

```
1. Item one

2. Item two

3. Item three
```

With an explicit newline between each point and then running the filter the output is:

```
1. Item one

1. Item two

1. Item three
```

**What is the expected behavior?**

I think the output should be:

```
1. Item one

2. Item two

3. Item three
```

Either reformatted as a list or as plain text but still preserving the correct numbers.

**Which versions of the filters or of Draft.js, and which browser / OS are affected by this issue? Did this work in previous versions of the filters / Draft.js?**

I cloned the repo and edited `FilterableEditor.js` to have a new on-demand filter button:

```js
<button onMouseDown={this.onFilter}>Filter</button>
```

With the filter as follows:

```js
  onFilter = () => {
    const { filtered, extended } = this.props
    const filters = {
      blocks: Object.keys(extended ? BLOCKS_EXTENDED : BLOCKS),
      styles: Object.keys(extended ? STYLES_EXTENDED : STYLES),
      entities: ENTITIES,
      maxNesting: extended ? MAX_NESTING_EXTENDED : MAX_NESTING,
      whitespacedCharacters: ["\n", "\t", "📷"],
    }

    const filteredState = filterEditorState(filters, this.state.editorState)
    this.setState({ editorState: filteredState })
  }
```

PS. Great stuff with this package - its super helpful :)

**Comment by thibaudcolas** [comment]
Hey @tomkelsey, thanks for reporting this! Glad you find this package helpful 🙂.

I think what’s happening here is that the filtering does not recognise this as a single list because of the newlines, but it might be easier to understand if you can share the `contentState` of the editor here, so we can see what blocks are there before and after filtering.

Could you also clarify what you mean by "newline"? Whether it’s creating a new block in the editor, or whether it’s just inserting a line break (`\n`).

**Comment by tomkelsey** [comment]
Hey thanks for getting back to me :)

I guess the easiest example I've just thought of is if you input:

`2. Hello`

```
{
  "blocks": [
    {
      "key": "4km2d",
      "text": "2. Hello",
      "type": "unstyled",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    }
  ],
  "entityMap": {}
}
```
Into an editor and run the filter the output will be:

`1. Hello`

```
{
  "blocks": [
    {
      "key": "4km2d",
      "text": "Hello",
      "type": "ordered-list-item",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    }
  ],
  "entityMap": {}
}
```

When I guess I'd expect it to just be `2. Hello` (and not converted into a list?)

---

For my original example:

```
1. Hello

2. Hello

3. Hello
```

```
{
  "blocks": [
    {
      "key": "4km2d",
      "text": "1. Hello",
      "type": "unstyled",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    },
    {
      "key": "c4os9",
      "text": "",
      "type": "unstyled",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    },
    {
      "key": "ecrei",
      "text": "2. Hello",
      "type": "unstyled",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    },
    {
      "key": "b4olh",
      "text": "",
      "type": "unstyled",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    },
    {
      "key": "nv84",
      "text": "3. Hello",
      "type": "unstyled",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    }
  ],
  "entityMap": {}
}
```

Is converted to:

```
1. Hello

1. Hello

1. Hello
```

```{
  "blocks": [
    {
      "key": "4km2d",
      "text": "Hello",
      "type": "ordered-list-item",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    },
    {
      "key": "c4os9",
      "text": "",
      "type": "unstyled",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    },
    {
      "key": "ecrei",
      "text": "Hello",
      "type": "ordered-list-item",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    },
    {
      "key": "b4olh",
      "text": "",
      "type": "unstyled",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    },
    {
      "key": "nv84",
      "text": "Hello",
      "type": "ordered-list-item",
      "depth": 0,
      "inlineStyleRanges": [],
      "entityRanges": [],
      "data": {}
    }
  ],
  "entityMap": {}
}
```

Perhaps a solution (if possible) would be to check that the number in the first bullet point is 1 before converting to an ordered list? If greater than 1 leave as plain text?

I guess that would cause my second example to still be wrong - the first item would be converted to a list but the subsequent items would be plain text.

Perhaps you could also check that there is more than one item before converting to a list? That way my second example would all stay as plain text

**Comment by thibaudcolas** [comment]
With your extra description I think it’s quite clear to me that what you are running against is intentional. Those blocks are converted to list items because they are prefixed with something the filters recognise as a list prefix, and that conversion is needed to retain ordered lists from Word, Google Docs and Dropbox Paper.

It also makes sense that it’s creating list items all numbered "1.", since all the list items in the input are separated by unstyled blocks (not newlines).

---

To solve your problem, I think the easiest you could do would be to redefine your own `filterEditorState` based on the other filters. You can then either disable all transformations of blocks by text prefixes, or just the ones that are currently problematic:

https://github.com/thibaudcolas/draftjs-filters/blob/2f5294bb18fc0deca48ae13f3d7be98dc2fd66a5/src/lib/filters/editor.js#L89-L149

Everything in there is meant to be reusable and composable, except for `PREFIX_RULES`.

---

Anyway, I’ll see if I can think of a good way to make this behavior configurable without resorting to defining your own `filterEditorState`, but I’m not sure what this will look like just yet.

**Comment by thibaudcolas** [comment]
@tomkelsey I’ve just pushed a change which should be released shortly, and allow you to disable or customise this behavior,

```diff
  onFilter = () => {
    const { filtered, extended } = this.props
    const filters = {
      blocks: Object.keys(extended ? BLOCKS_EXTENDED : BLOCKS),
      styles: Object.keys(extended ? STYLES_EXTENDED : STYLES),
      entities: ENTITIES,
      maxNesting: extended ? MAX_NESTING_EXTENDED : MAX_NESTING,
      whitespacedCharacters: ["\n", "\t", "📷"],
+      blockTextRules: [],
    }

    const filteredState = filterEditorState(filters, this.state.editorState)
    this.setState({ editorState: filteredState })
  }
```

Note setting `blockTextRules` to an empty array as I did here will completely disable conversion of blocks based on their text – if you want to keep some of the built-in conversions, you’ll need to copy-paste them as needed from https://github.com/thibaudcolas/draftjs-filters/blob/2850d7219abe8f17411c93376c62448accd1516d/src/lib/filters/editor.js#L65-L98.

**Comment by thibaudcolas** [comment]
:tada: This issue is fixed in v2.3.0, available on npm: [draftjs-filters@2.3.0](https://www.npmjs.com/package/draftjs-filters).

<p align="right">
Generated by :package::rocket: <a href="https://github.com/semantic-release/semantic-release">semantic-release</a>
</p>

**Comment by tomkelsey** [comment]
Many thanks @thibaudcolas :) Will have a look at this next week hopefully!
